### DISCUSSION TITLE
[Shutters][Rule] automatic calibration

### DISCUSSION THREAD

**Discussion by Gifford47** [arendst/Tasmota#15050] [Q&A]
Hi!
i want to write a rule, that measures the max current, the closing and opening time when opening and closing the shutter. i have several shelly2.5 modules that have to be configured and i want to do it automatically.

i have written the following rule, but i do not know how i measure the time until the shutter opened or closed:
it's the line with: ```var3=[180-CURRENT_RULETIMER1_VAL]+1```

```
var1 -> calibration flag
var2 -> max current on shutter close
var3 -> time till shutter closes
var4 -> max current on shutter open
var5 -> time till shutter opens


rule2
    on EVENT#calibrate do backlog var1=1; var2=0; var4=0; shutterclose; ruletimer1 180 endon
    on Energy#Current[0]>var2 do backlog if(%var1%==1) var2=%value% endif endon
    on Energy#Current[0]=0 backlog if(%var1%==1) var3=[180-CURRENT_RULETIMER1_VAL]+1; shutteropen; ruletimer1 180 endif endon
    on Energy#Current[1]>var4 do backlog if(%var1%==1) var4=%value% endif endon
    on Energy#Current[1]=0 backlog if(%var1%==1) var5=[180-CURRENT_RULETIMER1_VAL]+1; event calibrate_end endif endon
    on EVENT#calibrate_end do backlog ShutterCloseDuration %var3%; ShutterOpenDuration %var5%; ShutterMotorDelay 1; rule1 "on energy#current[0]>%var2% do shutterstop endon on energy#current[1]>%var4% do shutterstop endon"; rule1 5; rule1 1
    
rule2 1 
```

**Comment by Gifford47** [comment]
Any suggestions how i calculate a time difference?

**Reply by sfromis** [reply]
For 1-second resolution, you can issue the `RuleTimer` command for a "too long" interval, and again without value to query remaining time, using a JSON-based trigger to pick up the value. With a subtraction you then have a time difference.

If you want 0.1-second resolution, you can do similar tricks with a dummy relay pin, and `PulseTime`, but that's a bit more messy to work with, as it involves configuring another Power channel.

Note that using energy monitoring to detect timing has a rather low accuracy for precise timing.

**Reply by Gifford47** [reply]
but is there an option without another rule (on ... endon) to read the remaining time after executing "RuleTimer1" command ?

**Reply by Gifford47** [reply]
> For 1-second resolution, you can issue the `RuleTimer` command for a "too long" interval, and again without value to query remaining time, using a JSON-based trigger to pick up the value. With a subtraction you then have a time difference.
> 
> If you want 0.1-second resolution, you can do similar tricks with a dummy relay pin, and `PulseTime`, but that's a bit more messy to work with, as it involves configuring another Power channel.
> 
> Note that using energy monitoring to detect timing has a rather low accuracy for precise timing.

but how can i trigger at "T1" when the response after executing "RuleTimer" is as follows: ```{"T1":152,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}```?

at the moment i have:
```
on T1 do
	var3 %value%
endon
```
but it doesent work...

**Reply by sfromis** [reply]
Your rule there is ok, but if you execute the RuleTimer command from a rule, you should have it in a Backlog, also if it is the only command in the rule.

**Reply by Gifford47** [reply]
> Your rule there is ok, but if you execute the RuleTimer command from a rule, you should have it in a Backlog, also if it is the only command in the rule.

i have a backlog command, [see](https://github.com/arendst/Tasmota/discussions/15050#discussioncomment-2337153):
```
on energy#current[0]=0 do
	backlog if(%var1%==1) RuleTimer; var4=180-var3+1; shutteropen; ruletimer1 180 endif
endon 
```

**Reply by sfromis** [reply]
Works for me, in a small test case essentially matching what you do:
```
rule1 on T1 do var1 T1=%value% endon on Event#test do backlog if (%value%=1) RuleTimer endif endon
10:35:19.288 CMD: event test=1
10:35:19.295 MQT: xxxxxxxx/stat/EVENT = {"Event":"Done"}
10:35:19.330 RUL: EVENT#TEST performs "backlog if (1=1) RuleTimer endif"
10:35:19.344 MQT: xxxxxxxx/stat/IF = {"If":"Done"}
10:35:19.594 MQT: xxxxxxxx/stat/RULETIMER = {"T1":0,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
10:35:19.605 RUL: T1 performs "var1 T1=0"
10:35:19.612 MQT: xxxxxxxx/stat/VAR = {"Var1":"T1=0"}
```
Log shows details of the steps to get there....

**Reply by Gifford47** [reply]
ok, now that part works. but if i have the following rule, var2 and var5 have always the same value no matter which direction the roller shutter moves.

```
on energy#current[0]>var5 do
	backlog if(%var1%==1) var5=%value% endif
endon 
on energy#current[1]>var2 do
	backlog if(%var1%==1) var2=%value% endif
endon
```

**Reply by sfromis** [reply]
In the trigger part of a rule, use `%var5%` to resolve the value.

**Reply by Gifford47** [reply]
thank you, I should have seen that too...
nevertheless, the rule still does not work. when i close the shutter, it triggers nothing. when i open it, var2 AND var5 are set to the same value:

```
on energy#current[1]>%var5% do
	backlog var5 %value%
endon 
on energy#current[0]>%var2% do
	backlog var2 %value%
endon
```

**Reply by sfromis** [reply]
You did not get my hint about logs? There, you can follow that happens

**Comment by Gifford47** [comment]
ok, i will now test the following rule:

```
Module: Shelly2.5
#define USE_EXPRESSION         // Add support for expression evaluation in rules (+3k2 code, +64 bytes mem)  
#define SUPPORT_IF_STATEMENT   // Add support for IF statement in rules (+4k2 code, -332 bytes mem)
Template: {"NAME":"Shelly 2.5","GPIO":[56,0,17,0,21,83,0,0,6,82,5,22,156],"FLAG":2,"BASE":18}

var1 -> calibration flag
var2 -> max current on shutter close
var3 -> ruletimer1 remaining time
var4 -> time till shutter closes
var5 -> max current on shutter open
var6 -> time till shutter opens


rule2

on EVENT#calibrate do
	backlog var1 1; var2 0; var5 0; shutterclose; ruletimer1 180
endon 
on T1 do
	var3 %value%
endon 
on energy#current[0]>var2 do
	backlog if(%var1%==1) var2 %value% endif
endon 
on energy#current[0]=0 do
	backlog if(%var1%==1) RuleTimer; var4=180-var3+1; shutteropen; ruletimer1 180 endif
endon 
on energy#current[1]>var5 do
	backlog if(%var1%==1) var5 %value% endif
endon 
on energy#current[1]=0 do
	backlog if(%var1%==1) RuleTimer; var6=180-var3+1; event calibrate_end endif
endon
    on EVENT#calibrate_end do backlog ShutterCloseDuration var4; ShutterOpenDuration var6; ShutterMotorDelay 1; rule1 "on energy#current[0]>%var2% do shutterstop endon on energy#current[1]>%var5% do shutterstop endon"; rule1 5; rule1 1
    
rule2 1
```

**Comment by Gifford47** [comment]
how i trigger the current ruletimer1 value?
i tried Rules#Timer or Time#Second to get the current seconds, but nothing works/ triggers ...

**Reply by sfromis** [reply]
When you issue the RuleTimer command without arguments, the JSON response appearing will have the remaining intervals. You can use a JSON-based trigger to pick up values, by exactly matching the object and field name used.

**Comment by Bascht74** [comment]
@Gifford47: As I am switching from ESP8266 to ESP32 that script would help me. Did you tweak it more or is this (https://github.com/arendst/Tasmota/discussions/15050#discussioncomment-2337153 )  the "final" version that I could use?

**Reply by Gifford47** [reply]
Yes, that's the last revision but it doesn't work at the moment because the timer is not saved to the var. If you can please improve it and share it here.

**Comment by MichaelReuter84** [comment]
Hi,
following steppes and code worked on my side:

**Preperation:**

Start TasmoCompiler in docker:
	sudo docker pull benzino77/tasmocompiler
	sudo docker run --rm --name tasmocompiler -p 3000:3000 benzino77/tasmocompiler
Then point your browser to 
	http://localhost:3000

**Compiling settings for Tasmota for Shelly2.5 ESP8266:**
```
	#define WEBSERVER_ADVERTISE
	#undef MQTT_HOST_DISCOVERY   
	#define MDNS_ENABLED           1
	#undef USE_ADC_VCC
	#define USE_ADE7953
	#define USE_SHUTTER
	#define USE_EXPRESSION         // Add support for expression evaluation in rules (+3k2 code, +64 bytes mem)  
	#define SUPPORT_IF_STATEMENT   // Add support for IF statement in rules (+4k2 code, -332 bytes mem)
```

**Template:** 	
`{"NAME":"Shelly 2.5 (buttons)","GPIO":[320,0,34,0,224,129,0,0,640,128,608,225,3456,4736],"FLAG":0,"BASE":18}`
mportant! After setting the template, a hardware reset is necessary so that everything is displayed correctly! (If the Shellys are already installed in the box, the fuse in the fuse box can also be used)
					

shutter have to be fully open (use "POWER1 1" or do it via button )...

Determine values for power calibration. (For more precise calibration, the Shelly must be calibrated "at the table" with a defined consumer...) :
	PowerSet 100 (100 Watt out of datasheet) 
	VoltageSet 220 (220 Volt measured with Multi-meter before, e.g @socket) 
        CurrentSet 455        <-- 1000 * (100.0/220) = 455 mA
  
var1 -> time till shutter opens
var2 -> time till shutter closes
var3 -> ruletimer1 remaining time
var4 -> set threshold to trigger current check 0 Position
var5 -> set threshold to trigger current check 0 Position
var6 -> default Duration for Shuttermovment in sec
mem1 -> current on shutter open
mem2 -> current on shutter closes
mem11 -> mem1 *1,40 (+40% for overload protection)
mem12 -> mem2 *1,40 (+40% for overload protection)
mem3 -> Calibration Status (0 ... 6)
					
**Copy each rule to console activate rule1 and run event SetUpShutter:**
														

```
rule1
  ON Event#SetUpShutter DO Backlog rule2 OFF; rule3 OFF; SetOption80 1; SetOption21 1; Interlock 1; ShutterRelay1 1; mem3 0; restart 1 ENDON
  ON System#Boot Do IF(%mem3%==0) Backlog var1 0; var2 0.1; var4 0.1; var5 0.1; var6 100; mem1 0; mem2 0; delay 200; event ButtonMultipressMode ENDIF ENDON
  ON EVENT#ButtonMultipressMode DO Backlog SetOption1 1; SetOption11 1; SetOption32 20; ShutterButton1 1 up 1; ShutterButton1 2 down 1; delay 10; event CaliInit ENDON
  ON EVENT#CaliInit DO Backlog ShutterSetOpen; ShutterCloseDuration %var6%; ShutterOpenDuration %var6%; delay 10; event CaliInit2 ENDON
  ON EVENT#CaliInit2 DO Backlog ruletimer1 %var6%; POWER2 1; rule2 ON ENDON
  ON T1 DO var3 %value% ENDON	
```

						

```
rule2
  ON Event#CaliPower2 DO Backlog PowerSet2 100; VoltageSet 220; CurrentSet2 455; mem3 2 ENDON
  ON ENERGY#current[2]>%mem2% DO IF(%mem3%==0) Backlog mem3 1; Event CaliPower2 ELSEIF(%mem3%==2) mem2 %value% ENDIF ENDON 
  ON ENERGY#current[2]<%var4% DO IF(%mem3%==2) Backlog mem3 3; var4 -1; RuleTimer; shutterstop; ShutterSetClose; event CaliDownDone ENDIF ENDON
  ON Event#CaliDownDone DO Backlog mem12=%mem2%*1.4; var2=%var6%-%var3%; rule2 OFF; delay 10; ruletimer1 %var6%; POWER1 1; rule3 ON BREAK
```
  
```
rule3
  ON Event#CaliPower1 DO Backlog PowerSet1 100; CurrentSet1 455; mem3 5 ENDON
  ON ENERGY#current[1]>%mem1% DO IF(%mem3%==3) Backlog mem3 4; Event CaliPower1 ELSEIF(%mem3%==5) mem1 %value% ENDIF ENDON 
  ON ENERGY#current[1]<%var5% DO IF(%mem3%==5) Backlog mem3 6; RuleTimer; var5 -1; shutterstop; ShutterSetOpen; event CaliUpDone ENDIF ENDON
  ON Event#CaliUpDone DO Backlog mem3 6; mem11=%mem1%*1.4; var1=%var6%-%var3%; event CaliFinish ENDON
  ON Event#CaliFinish DO Backlog ShutterCloseDuration %var2%; ShutterOpenDuration %var1%; ShutterSetHalfway 67; shuttermotorstop 300; ShutterMotorDelay 0.1; event MoveCheck ENDON
  ON Event#MoveCheck DO Backlog delay 20;shutterposition 50; delay %var1%*500 ;rule3 OFF BREAK
```

To run the calibration:
  Backlog rule1 ON; event SetUpShutter

To check the values:
  backlog ShutterCloseDuration; ShutterOpenDuration; mem


**after calibratin (e.g. for CCU or RaspberryMatic controll):**

```
rule1
  ON Shutter1#Target DO Backlog Var2 %value%; Var1 0 BREAK
  ON ENERGY#current[1]>0.1 DO IF(%value%>%mem11%) backlog shutterstop ELSE Var1 %value% ENDIF ENDON
  ON ENERGY#current[2]>0.1 DO IF(%value%>%mem12%) backlog shutterstop ELSE Var1 %value% ENDIF ENDON
  ON SHUTTER1#Position DO IF(%Var1%>0.1) Backlog var3 0; WebSend [192.168.99.5:8181] /cuxd.exe?ret=dom.GetObject('CUxD.CUX2801002:11.SET_STATE').State(-%value%) ELSEIF(%var3%==0) delay 7; var3 1 ELSE shutterposition %var2% ENDIF ENDON
  ON IPAddress1 DO tmsend %value% ENDON
  ON System#Boot DO Backlog IPAddress; ENDON
  ON Tele-ANALOG#Temperature do WebSend [192.168.99.5:8181] /cuxd.exe?ret=dom.GetObject('TempInnenShellySchlafzimmer').State(%value%) ENDON
```
